Adds lensing lightcone computation using the Born approximation

EiffL opened the pull request:
Thanks to fantastic work by @dlanzieri we are in the position of opening this PR which aims to solve #45 .It currently implements a tool for computing lightcones and exporting lens planes, as well as a Born approximation of ray tracing.

Here are the goals before turning it into a full Pull Request:
- [x] : Add unit tests for lightcone and lensplane computation, probably against fastpm
- [x] : Add unit tests for lensing kernel and intermediary functions if possible
- [x] : Add unit test to valide the lensing power sprectrum

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,236 @@
+     2D mesh area in rad^2 per pixel
+     
+    """
+    return ((5*np.pi/180/nc_xy)**2)
```
if I understand correcly, here the 5 is for the 5 deg of the simulated field, we should probably leave that as a parameter of the function

EiffL commented on flowpm/lenstf.py:
```diff
@@ -0,0 +1,98 @@
+        z = tfp.math.interp_regular_1d_grid(d, 1e-12, 1.5e3, tf.convert_to_tensor(chis, dtype='float'), name='interpolation')
+        return z
+
+def wlen(d, ds, cosmo, boxsize, boxsize2D, mesh2D, name='efficiency_kernel'):
```
is this function now duplicated with part of the raytracing.py file ?

EiffL commented on flowpm/lenstf.py:
```diff
@@ -0,0 +1,98 @@
+        xy = tf.gather(y, (0, 1), axis=1, name='gather_xy')
+        return xy, d
+
+def z_chi(d, cosmo, name='z_chi'):
```
this is probably no longer needed since we have tensorflow code for that

EiffL reviewed (commented):
looking good already :-) I made a few comments

EiffL commented on flowpm/scipy/interpolate.py:
```diff
@@ -38,6 +38,6 @@ def interp_tf(x, xp, fp):
     xp0= tf.gather(xp, ind)
     xp1= tf.gather(xp, ind + tf.cast(tf.sign(s),dtype=tf.int64)) - xp0
-    a = (fp1)/(xp1)
+    a = (fp1)/(xp1+1e5)
```
you meant 1e-5

dlanzieri reviewed (commented):
thanks @EiffL! With these latest changes we should have solved all the technical problems we had observed, now we just have to test and validate it!

EiffL commented:
I've had a little time to look at this lenstools problem ^^' after struggling a bit, I figured out the correct way to export our lensplanes and made a small notebook here:
https://github.com/modichirag/flowpm/blob/lensing/notebooks/dev/test_lenstools.ipynb

But, in doing so, I realized that most of the lensing efficiency kernel computation needed to turn simulation volumes to lens planes happens outside of the raytracing functions in lenstools.
This is still a pretty cool check to to, especially when we will look at other ray tracing strategy, but we'll need another test that starts from the particle cube itself to validate the rest of the lensing efficiency computation.

EiffL commented:
After a little bit of exploration, got a mostly working direct comparison between LensTools and our own raytracing:

![image](https://user-images.githubusercontent.com/861591/112737766-380c6900-8f5d-11eb-97b5-afd9c6dd1b13.png)

Added a notebook with temporary demo code here:
https://github.com/DifferentiableUniverseInitiative/flowpm/blob/bf6d45309fc6c316907e7cf28ca5c20c5ae1d16e/notebooks/dev/lenstools_lightcone_test-adapted.ipynb

It's looking mostly comparable, LensTools applies a smoothing to the lensplanes so I kind of expect the map to be smoother, but   I'm not sure the projection matches 100%.
Must investigate more....

dlanzieri reviewed (approved):
Thank you for your work @EiffL! 
However, I want to be sure I've got it right. 
 So, from the original Lightcone function you are exporting the lens plans and the snapshots.
Then, instead of testing the raytracing  using these lens plans (as our previous test), you use the snapshots to create the lens plans with Lenstools, Compute the Born Approximation with them and compare them with our results .

EiffL commented:
That's exactly right. so, this should test the entire ray tracing procedure. The results don't match 100% because there is a fundamental difference in how we build the lens planes compared to lenstools.
LensTools (and as far as I can see most/all ray tracing codes) build the lensplanes in comoving coordinates, then interpolate the lensplanes on the angular grid on the sky during the ray tracing.
In what we do, which is inspired from MADLens, we instead directly build the lensplanes in the angular grid on the sky. 

Where it differs, is that the lenstools procedure imposes a maximal resolution in comoving scales, whereas our approach imposes a resolution in angular scales. And that makes a difference since different comoving scales correspond to different angular scales at different redshifts.
So for instance, here is how the same lensplanes appear under the 2 procedures:
![image](https://user-images.githubusercontent.com/861591/112752148-e5b95f80-8fd1-11eb-968e-43d0fc979afb.png)
![image](https://user-images.githubusercontent.com/861591/112752160-ec47d700-8fd1-11eb-847b-41460b2ed66a.png)

also, there seems to be a small difference in projection, still trying to understand where it comes from.

I've thought about it..... and I think it may be safer to follow the LensTools approach instead of what we are doing...

dlanzieri commented:
@EiffL ok, I see. So probably in this case creating a test code makes no sense because the difference in how we build the lens planes will make test fail.
So we can implement a new lightcone (or at least changing part of it) following the LensTools approach.

EiffL commented:
Ok, I'm running out of time for now and didn't get a chance to write a full function :'-( , but I added a demo notebook that computes lensplanes like lenstools:
https://github.com/DifferentiableUniverseInitiative/flowpm/blob/lensing/notebooks/dev/lenstools_lightcone_test-adapted-v2.ipynb
The lightcone function there exports density planes in comoving coordinates, similar to what this LensTools function does:
https://github.com/apetri/LensTools/blob/9151988bfe6fbd6809353a33cfb556d44b6806ed/lenstools/simulations/nbody.py#L522

Then I compare the density planes to the ones lentools computes
![image](https://user-images.githubusercontent.com/861591/112757340-bf072300-8fe9-11eb-98f8-806dbd5eebd5.png)
(left to right, lenstools, flowpm, difference)
 and it looks pretty good. It's not 100% the same because we use CIC painting and lenstools does nearest neighbor painting apparently, or something like that. But close enough that we can compare the 2, with an appropriate margin for error.

Then at the end of the notebook, I show how to use the `tfa.image.interpolate_bilinear` function of tensorflow-addons to compute the projected field on the sky.
What remain to do is simply to write the born approximation function equivalent to 
https://github.com/apetri/LensTools/blob/9151988bfe6fbd6809353a33cfb556d44b6806ed/lenstools/simulations/raytracing.py#L1332

dlanzieri commented:
@EiffL ok, thank you! Maybe I need a bit of time to get a handle on this, especially if I want to write the Born Approximation.

EiffL commented:
Sorry it took me a few hours to document what I did >.< I made some edits that are messing up the notebook, so instead of pushing my new version of it, I'll share here the steps that I followed so that you can reproduce them. There are essentially 2 things:

1. It appears that the convention for center of voxels is offset by half a voxel between the painting done in flowpm and lenstools for the lensplanes, which explains why we look carefully at the residual images, we see patterns like this:
![image](https://user-images.githubusercontent.com/861591/113058242-371e4600-91ae-11eb-96fe-5e3c6450114a.png)
which seem to be perpendicular to the direction of the first bissector (y=x).

In any case, I see this pattern in the residual image disappear if I do the following modification in the customized `FastPMSnapshot.getPositions()` function included in the notebook:
```python
    #Read in positions in Mpc/h
    if (first is None) or (last is None):
	    positions = (data["0/Position"][:] + 0.5/64*200 )*self.Mpc_over_h
    else:
	    positions = data["0/Position"][first:last]*self.Mpc_over_h
``` 
See I added a shift of `0.5/64*200` which corresponds to a distance of half a voxel, converted to Mpc/h. So we should just make that shift a little bit more generic, i.e. using the values included in `self.header` but otherwise I think this is as good a place as any to have this.

2. The second thing I did is, as we discussed, applied a little bit of smoothing to the lensplanes before computing the residuals, to remove the fine details of the painting procedure. for this I use the following:
```python
from scipy.ndimage import gaussian_filter
# And for instance to get the smoothed lensplane
smoot_ps = gaussian_filter(ps[ind],1)
```
and I get this:
![image](https://user-images.githubusercontent.com/861591/113059624-f4f60400-91af-11eb-8c85-aba7a54052da.png)
which looks very good :-)

EiffL commented:
didn't want to overcrowd our dev folder with notebooks, so I put the notebook I used for the plots above in a separate gist :-) https://gist.github.com/EiffL/65fc557ab487a9eba4d4abfcb9f1f266
Just in case it might be useful to look at the whole thing to understand what I did

EiffL commented:
Let me make a TODO list:

- [ ] : Implement this new lightcone function in FlowPM in the lensing branch (it exports lensplanes in comoving coordinates)
- [ ] : Validate the density planes themselves -> adding a test that compares FlowPM density plane with a lenstools density plane.  A new test file `test_raytracing.py`, with inside the FastPM class to reload snapshots
- [ ] : A test that just validates the interpolation from comoving distance to angular coordinates
- [ ] : A test that checks the Born approximation ray tracing, this can be an end-to-end test from the snapshots, or more simple, just testing the lenstool function from the same density planes in FlowPM and lenstools.

EiffL commented:
@dlanzieri so with this last update, I think we are almost all good to merge this branch ;-) We would just like to see now if we can make a simulation with a reasonable lensing power spectrum. We just may not be able to exactly validate it against theory, but I guess that's life.

EiffL commented:
Hi @dlanzieri, I tried something that we talked about today to reduce how much memory we need, and make a lightcone reusing the simulation volume several times. Have a look here :-) :
https://github.com/DifferentiableUniverseInitiative/flowpm/blob/lensing/notebooks/dev/test_lensing_random.ipynb

And I can make pretty nice convergence maps:
![image](https://user-images.githubusercontent.com/861591/115468565-8c8cc680-a233-11eb-868e-0ed06ea27ae6.png)

But there is something wonky going on when I compare the power spectrum to the theory.... it doesnt match for all source redshifts, not sure why....
It works for some source redshifts though:
![image](https://user-images.githubusercontent.com/861591/115468711-c362dc80-a233-11eb-91b0-f4e65486f187.png)
Might be a problem with the fact that we only have coarse lens planes here.... not sure

dlanzieri commented:
Hi @EiffL ! I see, I tried to get the same plot and also to plot the power spectrum for different source redshifts.
Actually, the power spectrum obtained from the convergence map seems unlikely to change with different source redshifts.
![compps](https://user-images.githubusercontent.com/72620117/115538389-1cfef180-a29c-11eb-90fe-924a2bbaa1ad.png)

EiffL commented:
Interesting! Hummmm.... Can you plot the flowpm power spectra for all source redshifts ? I wonder what it looks like for smaller redshifts

dlanzieri commented:
ok, scratch what I just said, the Power spectrum changes with different source redshifts
<img width="755" alt="Schermata 2021-04-21 alle 15 14 07" src="https://user-images.githubusercontent.com/72620117/115559750-4592e580-a2b4-11eb-9962-d7054e41bac7.png">

EiffL commented:
@dlanzieri I think we are in good shape with this PR :-) 
There is one last thing that we could try to do before merging.... based on our latest results, it looks like an approach that reuses the simulation volume several times with random shifts/rotations is a good idea as it would save a lot of memory. So, we could add the tools needed for that to `raytracing.py`.

Do you want to take a look at the code that @Maxelee already added here: https://github.com/DifferentiableUniverseInitiative/flowpm/blob/d8d9036854cefc7b1cbff7a2ca739eae91732eb4/flowpm/experimental/lenstf.py#L8
and here:
https://github.com/DifferentiableUniverseInitiative/flowpm/blob/d8d9036854cefc7b1cbff7a2ca739eae91732eb4/flowpm/experimental/lenstf.py#L72

and see if we can reuse pieces of this code to do a smarter thing than what I hacked here:
```python
    plane = tf.roll(plane, shift=np.random.randint(0,256), axis=1)
    plane = tf.roll(plane, shift=np.random.randint(0,256), axis=2)
    plane = tf.expand_dims(plane, axis=-1)
    plane = tf.image.random_flip_left_right(plane)
    plane = tf.image.random_flip_up_down(plane)
```
https://github.com/DifferentiableUniverseInitiative/flowpm/blob/lensing/notebooks/dev/test_lensing_random.ipynb

Do you want to take a stab at it?

dlanzieri commented:
Hi @EiffL ! Of course, Let's do it :)

EiffL commented:
Great :-) So, you can try to hack a prototype first, but ultimately, I think we want to modify the `density_plane` function 
https://github.com/DifferentiableUniverseInitiative/flowpm/blob/d8d9036854cefc7b1cbff7a2ca739eae91732eb4/flowpm/raytracing.py#L15
to add an option to apply a given rotation to the density cube, and a given shift, when cuting the density plane.

EiffL commented on flowpm/__init__.py:
```diff
@@ -1,3 +1,6 @@
 import flowpm.cosmology as cosmology
+import flowpm.tfbackground as background
+import flowpm.io as io
+import flowpm.raytracing as raytracing
```
```suggestion
import flowpm.raytracing as raytracing

```

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+
+
+def generate_matrix():
+  """ Generate the possible rotation matrices
```
Add docstring docucmentation using the formating convention we adopted. See here https://www.tensorflow.org/community/contribute/code_style#tensorflow_conventions_and_special_uses

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+import flowpm.constants as constants
+
+
+def generate_matrix():
```
```suggestion
def rotation_matrices():
```

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+  x = np.asarray([1, 0, 0], dtype=int)
+  y = np.asarray([0, 1, 0], dtype=int)
+  z = np.asarray([0, 0, 1], dtype=int)
+  M_matrices = [np.asarray([x,y,z],dtype=int), np.asarray([x,z,y],dtype=int),np.asarray([z,y,x],dtype=int),np.asarray([z,x,y],dtype=int), \
```
the matrices will be used to apply rotations on float tensors, so we probably want to use a dtype of float32 instead of int

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+  return M_matrices
+
+
+def shift():
```
```suggestion
def random_shift():
```

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+
+
+def shift():
+  """ Generate possible xy shifts
```
The documentation doesn't tell me that this function generates random 2d shifts

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+  I = np.zeros(3)
+  facx = np.random.uniform(0, 1)
+  facy = np.random.uniform(0, 1)
+  xyshifts = [np.asarray([facx, facy, 0], dtype=float)]
```
```suggestion
  xyshifts = [np.asarray([facx, facy, 0], dtype=np.float32)]
```

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+def density_plane(state,
+                  nc,
+                  center,
+                  Matrix,
```
Two things ^^'
1. we should rename the `Matrix` argument into `rotation` 
2. we should place the rotation and shift arguments at the end of the list, just before `name='...'`, and set them to `None` by default. The reason for moving them to the end, is because you don't want to separate thee `center` from the `width` arguments, together they define what the plane is going to look like

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+  """ Extract a slice from the input state vector and
+  project it as a density plane.
+
+  Args:
```
We need to add documentation for the new arguments

EiffL replied:
done

EiffL commented on flowpm/raytracing.py:
```diff
@@ -0,0 +1,142 @@
+
+    shape = tf.shape(pos)
+    batch_size = shape[0]
+    y = tf.einsum('ij,bkj->bki', Matrix, pos)
```
we will probably only want to do this if `rotation` and/or `shift` is not `None`

EiffL replied:
node

EiffL reviewed (commented):
Loooking good, but marked a few comments for minor changes

EiffL commented:
Ok, I've done a full Review of the code and fixed a couple of last issue. The lensing notebook now can produce these nice plots:
![image](https://user-images.githubusercontent.com/861591/116329468-b0b84c80-a7cb-11eb-8098-dbd9bf6d8eae.png)
for a cube of 128x128x128 particles, on a 5x5 deg field, and we kind of agreed it was not completely crazy for a low resolution simulattion. 
In any case all of our tests against lenstools seem to pass, so I think we are all good to go to merge this branch \o/

EiffL commented:
@dlanzieri since I opened the PR I can't write an approval review myself :-) Can you do a last review and see if it looks all good to merge?

EiffL commented:
also, note that in order to have a clean `notebooks` folder that doesn't scare away users ^^' I have moved any development notebook to the `notebooks/dev`  folder.

dlanzieri reviewed (approved):
@EiffL Thank you for this full review! That looks all right, I think we are ready to merge :)

EiffL commented:
Fantastic! Thank you @dlanzieri ! Merging \o/
